### PR TITLE
Add rake task to report lost payments

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Related to #3924 and #3925.

In rare cases, payment records can get lost through database rollbacks. This new rake task tries to find log entries for lost transactions and reports them as CSV. It can help us until the underlying bug is fixed. It will also help us monitor if we really fixed the bug or if this happens again.

I'm not sure if we should make this more sophisticaed (run through schedule.rb) or keep it as temporary tool we want to get rid of as soon as possible.

#### What should we test?

No testing required.

#### Release notes

Added script for sys admins to find lost payments.

Changelog Category: Added

### CODE REVIEW

**Comment by luisramos0** [comment]
you can extract method to make rubocop happier :-)

**Comment by mkllnk** [comment]
@luisramos0 Ah, yes. Rubocop tricked me there. We are missing a value for BlockLength in our style guide and the todo file overrode it. I think that it's actually a mistake in our config. But I also fixed the style in this commit.

By the way, a maximum method length of 10 doesn't work when you define lists within a method. The headers could go in a constant but what about the values? Do you have a better idea?

**Comment by luisramos0** [comment]
nice catch Maikel, I'd vote for keeping the limit of 10 and using rubocop exception in places where you need to a define long lists.

File: lib/tasks/missing_payments.rake [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# Report of the last 35 days:
+#   rake ofn:missing_payments[35]
+namespace :ofn do
+  desc 'Find payments that got lost'
+  task :missing_payments, [:days] => :environment do |_task_, args|
+    days = args[:days].andand.to_i || 7
+    payments_sequence = Spree::Payment.where("created_at > ?", days.days.ago).order(:id).pluck(:id)
+    payments_range = (payments_sequence.first..payments_sequence.last).to_a
+    missing_payment_ids = payments_range - payments_sequence
+    puts "Gaps in the payments sequence: #{missing_payment_ids.count}"
```
Comment (sauloperez) [inline lib/tasks/missing_payments.rake]: We should avoid using `puts` and use `Rails.logger` instead. The Rails logging is better suited for these.

**Comment by sauloperez** [comment]
> keep it as temporary tool we want to get rid of as soon as possible.

:+1:

File: lib/tasks/missing_payments.rake (hunk 1 above)
Comment (mkllnk) [reply lib/tasks/missing_payments.rake]: I don't see this as log message. It's the defined output of the program. It's like `ls` is listing files. I don't think it should use a logger for that.

**Comment by mkllnk** [comment]
@sauloperez 
> Do we know whether the log entry gets inserted within the lifespan of the transaction that rolls back?

It gets re-inserted after rollback:

https://github.com/openfoodfoundation/spree/blob/46d6f8f5fd434105b0c69958276d1727a3066a97/core/app/models/spree/log_entry.rb#L6-L14

File: lib/tasks/missing_payments.rake [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+# ## Usage
+
+# Report of the last 35 days:
+#   rake ofn:missing_payments[35]
+namespace :ofn do
+  desc 'Find payments that got lost'
+  task :missing_payments, [:days] => :environment do |_task_, args|
+    days = args[:days].andand.to_i || 7
+    payments_sequence = Spree::Payment.where("created_at > ?", days.days.ago).order(:id).pluck(:id)
+    payments_range = (payments_sequence.first..payments_sequence.last).to_a
```
[the remark this answers is not in the document]
Comment (mkllnk) [reply lib/tasks/missing_payments.rake]: You're right. Fixed in a new commit. Please re-review. :-)

**Comment by sauloperez** [comment]
There are at least 2 of those 3 Code Climate complaints that can be easily fixed.

**Comment by mkllnk** [comment]
@sauloperez I'm moving this to test ready.

**Comment by mkllnk** [comment]
Ah, no, it's no-test. ;-) I'm merging it then.

**Pull request merged**